contractcourt/chain_watcher.go

```diff
@@ -379,15 +379,45 @@ func (c *chainWatcher) closeObserver(spendNtfn *chainntnfs.SpendEvent) {
 					c.cfg.chanState.FundingOutpoint, err)
 			}
 
-		// This is the case that somehow the commitment
-		// broadcast is actually greater than even one beyond
-		// our best known state number. This should NEVER
-		// happen, but we'll log it in any case.
+		// This is the case that somehow the commitment broadcast is
+		// actually greater than even one beyond our best known state
+		// number. This should ONLY happen in case we experienced some
+		// sort of data loss.
 		case broadcastStateNum > remoteStateNum+1:
-			log.Errorf("Remote node broadcast state #%v, "+
+			log.Warnf("Remote node broadcast state #%v, "+
 				"which is more than 1 beyond best known "+
-				"state #%v!!!", broadcastStateNum,
-				remoteStateNum)
+				"state #%v!!! Attempting recovery...",
+				broadcastStateNum, remoteStateNum)
+
+			// If we are lucky, the remote peer sent us the correct
+			// commitment point during channel sync, such that we
+			// can sweep our funds.
+			commitPoint, err := c.cfg.chanState.DataLossCommitPoint()
```

Comment:
commintPoint -> commitPoint 

What happens if we detect this force close before we actually write this state to disk? Since this is usually expected _before_ we connect to the remote peer it seems likely. Also consider the case where they've already swept on-chain, but we weren't online, but they keep re-broadcasting the last chan sync message so we can get the revocation point. It would seem that we need some additional synchronization in order to resolve this in all scenarios.

Reply:
That is a very good point! 

I briefly started working on this, but realized that this require some bigger changes:

**1) In the case we are about to receive the message, but detecting the force close in chain wins the race.** We must retry fetching the commit point after a timeout, or a signal that a new commitpoint was received.

**2) In the case where the remote has closed the channel while we have been offline.** We must support resending the channel sync message such that we can continue when it is received.

Since 2) seems to require 1), it makes sense do add them both in the follow-up adding chanSync message resend. The current implementation will handle the case where we detect a desync during channel reestablishment, and one party force closes because of this (Note that 1) is unlikely in this case, as we wait for the force close to be confirmed before we act).

Reply:
SGTM